File: VB/NodeReordering/Models/DataHelper.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web

Namespace NodeReordering.Models
    Public Class DataHelper

        Private _sessionKey As String = "ExampleDataSource"
        Public Property Data() As List(Of SampleDataItem)
            Get
                If HttpContext.Current.Session(_sessionKey) Is Nothing Then
                    HttpContext.Current.Session(_sessionKey) = Me.GetSampleData()
                End If
                Return DirectCast(HttpContext.Current.Session(_sessionKey), List(Of SampleDataItem))
            End Get
            Set(ByVal value As List(Of SampleDataItem))
            End Set
        End Property

        Private Function GetSampleData() As List(Of SampleDataItem)
            Dim result As New List(Of SampleDataItem)()
            result.Add(New SampleDataItem("root", 0, Nothing))
            result.Add(New SampleDataItem("a", 1, 0))
            result.Add(New SampleDataItem("b", 2, 0))
            result.Add(New SampleDataItem("a1", 3, 1))
            result.Add(New SampleDataItem("a2", 4, 1))
            result.Add(New SampleDataItem("a3", 5, 1))
            result.Add(New SampleDataItem("b1", 6, 2))
            result.Add(New SampleDataItem("b2", 7, 2))
            result.Add(New SampleDataItem("b1a", 8, 6))
            result.Add(New SampleDataItem("b1b", 9, 6))
            result.Add(New SampleDataItem("b1c", 10, 6))
            Return result
        End Function

        Public Sub SwapDataItems(ByVal item1 As SampleDataItem, ByVal item2 As SampleDataItem)
            If item1 Is Nothing OrElse item2 Is Nothing Then
                Return
            End If
            Dim index1 As Integer = Data.IndexOf(item1)
            Dim index2 As Integer = Data.IndexOf(item2)
            Data(index1) = item2
            Data(index2) = item1
        End Sub

    End Class

    Public Class SampleDataItem
        Public Property Title() As String
        Public Property Key() As Integer
        Public Property ParentKey() As Integer?

        Public Sub New(ByVal title As String, ByVal key As Integer, ByVal parentKey? As Integer)
            Me.Title = title
            Me.Key = key
            Me.ParentKey = parentKey
        End Sub
    End Class
End Namespace

File: CS/NodeReordering/Models/DataHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;

namespace NodeReordering.Models
{
    public class DataHelper
    {

        private string _sessionKey = "ExampleDataSource";
        public List<SampleDataItem> Data
        {
            get
            {
                if (HttpContext.Current.Session[_sessionKey] == null)
                    HttpContext.Current.Session[_sessionKey] = this.GetSampleData();
                return (List<SampleDataItem>)HttpContext.Current.Session[_sessionKey];
            }
            set { }
        }

        private List<SampleDataItem> GetSampleData()
        {
            List<SampleDataItem> result = new List<SampleDataItem>();
            result.Add(new SampleDataItem("root", 0, null));
            result.Add(new SampleDataItem("a", 1, 0));
            result.Add(new SampleDataItem("b", 2, 0));
            result.Add(new SampleDataItem("a1", 3, 1));
            result.Add(new SampleDataItem("a2", 4, 1));
            result.Add(new SampleDataItem("a3", 5, 1));
            result.Add(new SampleDataItem("b1", 6, 2));
            result.Add(new SampleDataItem("b2", 7, 2));
            result.Add(new SampleDataItem("b1a", 8, 6));
            result.Add(new SampleDataItem("b1b", 9, 6));
            result.Add(new SampleDataItem("b1c", 10, 6));
            return result;
        }

        public void SwapDataItems(SampleDataItem item1, SampleDataItem item2)
        {
            if (item1 == null || item2 == null) return;
            int index1 = Data.IndexOf(item1);
            int index2 = Data.IndexOf(item2);
            Data[index1] = item2;
            Data[index2] = item1;
        }

    }

    public class SampleDataItem
    {
        public string Title { get; set; }
        public int Key { get; set; }
        public int? ParentKey { get; set; }

        public SampleDataItem(string title, int key, int? parentKey)
        {
            Title = title;
            Key = key;
            ParentKey = parentKey;
        }
    }
}